nomad/structs/structs.go

```diff
@@ -10527,20 +10569,12 @@ func (a *Allocation) RescheduleEligible(reschedulePolicy *ReschedulePolicy, fail
 	if reschedulePolicy == nil {
 		return false
 	}
-	attempts := reschedulePolicy.Attempts
-	enabled := attempts > 0 || reschedulePolicy.Unlimited
-	if !enabled {
-		return false
-	}
-	if reschedulePolicy.Unlimited {
-		return true
-	}
-	// Early return true if there are no attempts yet and the number of allowed attempts is > 0
-	if (a.RescheduleTracker == nil || len(a.RescheduleTracker.Events) == 0) && attempts > 0 {
-		return true
+
+	if a.RescheduleTracker == nil {
+		return reschedulePolicy.Attempts > 0
 	}
-	attempted, _ := a.rescheduleInfo(reschedulePolicy, failTime)
-	return attempted < attempts
+
+	return a.RescheduleTracker.RescheduleEligible(reschedulePolicy, failTime)
```

Comment:
By pushing all the checks _except_ the `ReschedulerTracker == nil` into this function, haven't we changed the order of checks such that we're returning based on the reschedule policy attempts even if `reschedulePolicy.Unlimited` is set? 

We could safely handle the `RescheduleTracker == nil` case at the top of the `RescheduleEligible` method (you can call a method with a pointer receiver even if the pointer is nil). That would let both the `Allocation` and `AllocListStub` version of this share their code entirely.

Reply:
> you can call a method with a pointer receiver even if the pointer is nil)

I was not aware of that, thanks for the tip! I just applied your suggestion, that solves everything: simpler and the confusion around the order of checks is gone